### ISSUE TITLE
Password reset via my-account/lost-password does not trigger after_password_reset action 

### ISSUE THREAD

**Issue by loranallensmith** [woocommerce/woocommerce#27795]
Howdy Core Team!  :wave:

I have been working with @jackmcconnell troubleshooting a compatibility issue between the [Better Notifications for WP](https://wordpress.org/plugins/bnfw/) plugin (BNFW) and WooCommerce.  One of the notifications that admins can set up will send a confirmation email to the user after a user's password is reset.  To do this, they're hooking to the [`after_password_reset`](https://github.com/WordPress/WordPress/blob/d4a319c4530bfb96977f3e416e84695f34eb55a8/wp-includes/user.php#L2495) action that fires when a WordPress user's password gets reset.

When a customer in WooCommerce uses the `my-account/lost-password` form to reset their password, the email notification indicating that the password has changed is not going out, however, those emails _do_ get triggered if an admin manually resets the user's password via wp-admin _or_ if the user changes their password via `my-account/edit-account/`

After digging into things a bit, I noticed that the issue seems to be stemming from the `reset_password` function that handles the password reset form in the `my-account/lost-password` flow. See below:

https://github.com/woocommerce/woocommerce/blob/02cf0dfaed5923513de0c88add597d1560c2cfd2/includes/shortcodes/class-wc-shortcode-my-account.php#L373-L382

It replicates some of the logic from WordPress' [`reset_password`](https://github.com/WordPress/WordPress/blob/d4a319c4530bfb96977f3e416e84695f34eb55a8/wp-includes/user.php#L2473-L2496) function, but it doesn't include the `after_password_reset` hook.

After chatting with @jackmcconnell about this, I figured we should open up an issue here in case this was a bug in that flow.  This behavior may be by design though.  Are y'all aware of any ill effects that could occur if we were to add that hook to the [`reset_password`](https://github.com/woocommerce/woocommerce/blob/02cf0dfaed5923513de0c88add597d1560c2cfd2/includes/shortcodes/class-wc-shortcode-my-account.php#L373-L382) function in WC?

**Comment by JVDL-1** [comment]
Hey Allen,

Did you find a solution for this problem yet? I'm having the same issue with another plugin which also uses the after_password_reset. Did inserting the hook directly in the WooCommerce file work for you?

Thank you.

**Comment by rrennick** [comment]
This just needs to have the hook added.

**Comment by loranallensmith** [comment]
Howdy core team. :wave:  I just wanted to follow up on this to see what other folks thought about the two solutions I proposed above:

> to swap out WooCommerce's direct call to `wp_password_change_notification` on line 380 with do_action('after_password_reset') instead. This is the action that WordPress has `wp_password_change_notification` hooked to already, so the net result would be the same (as far as the current interaction between WooCommerce and WordPress) 

or

> to provide a WC-specific action hook that coincides with Woo's custom `reset_password` function

I think both of these would let WC continue to respect the `woocommerce_disable_password_change_notification` filter while still allowing developers to run custom code at the end of the password reset process.

**Comment by mfs-mindsize** [comment]
@loranallensmith - I was scanning the "good first issue" issues and this one caught my eye :) 

`It replicates some of the logic from WordPress' reset_password function, but it doesn't include the after_password_reset hook.`

That said, to me, doesn't it makes the most sense to only use the WP reset, and then its hooks for any added WC particulars? Else, we get something like you found. Duplicated code that's close, not close enough, and that leads to unexpected results. 

Since there are no special WC hooks in its reset_password() there shouldn't be any backward compatibility issue, yes? Mind you, I've only scanned the threat once, and properly read it once. So perhaps I've missed something.

**Comment by loranallensmith** [comment]
Hey @mfs-mindsize!  :wave:  Thanks for taking a look at this as a potential contributor.  :tada:

From what I can tell, the filter online 379 of the WC function could cause some compatibility issues if it is circumvented at this point:
https://github.com/woocommerce/woocommerce/blob/f65491d1b21e8e28c4a26b20d7f36f8e28fd040f/includes/shortcodes/class-wc-shortcode-my-account.php#L379

Looks like this filter was added as a fix for #24153 to allow developers to programmatically disable the notification that WooCommerce sends out.  If WC delegated this call over to WP's `reset_password`, it would break functionality for developers who rely on that filter right now. That being said, delegating to WP's function gives those developers the option to disable the notification by removing the `wp_password_change_notification` action but it would introduce a breaking change to WC, which is something I know the WC Core team is aiming to avoid as much as possible.

FWIW, there also seem to be small code differences that are exclusive to each function:
- The WP function resets the option for the default password nag when it runs.
- The WC function sets a password reset cookie when it runs.

I don't know all the historical context around these two differences, so I can't say for sure how critical they are to the overall password reset flow in WooCommerce.

The approach I floated [above](https://github.com/woocommerce/woocommerce/issues/27795#issuecomment-721310464) of delegating to the `after_password_reset` hook is kind of a compromise.  It lets WC core maintain its contract with developers who rely on the `woocommerce_disable_password_change_notification` filter while still letting developers hook into the `after_password_reset` action in WP.

The big risk of this approach is that it turns the WC filter into a master switch for any functionality further down that call stack.  In other words, if one extension flips the switch to disable the functionality, any other extensions hooked into the action won't necessarily execute during the `my-account/lost-password` flow.

**Comment by mfs-mindsize** [comment]
@loranallensmith 

That filter? Yup. I noticed that :) But my theory is it can be addressed by using the hook in the WP reset_password()

`	do_action( 'after_password_reset', $user, $new_pass );`

That filter is after the reset so (I believe) it's a matter of hooking to the WP reset's action. 

Yes? No?

The only issue might be that once using that action, it can be removed. That said, the filter is optional anyway. So if the action was removed for some odd reason, it would be the same as passing false back in the filter.

**Comment by loranallensmith** [comment]
@mfs-mindsize I think I understand the flow you're describing.  Are you saying that WooCommerce could hook into WordPress'  `after_password_reset` action with a high-priority function that conditionally removes [the action that calls `wp_password_change_notification`](https://github.com/WordPress/WordPress/blob/d6cff7965a10c89e629a87eb6e4ecb0ed9614d50/wp-includes/default-filters.php#L458) based on the value of the WC filter?

> Then there's no need to carry a WC-specific reset. Yes? No?

I can't say with certainty.  Those other small differences in these functions may be significant, but I'm not familiar enough with the details of WordPress password resets in general to know whether or not that's the case.

Still, if you're looking for a good first issue, this could be a fun one.  It would be worth trying out the change you're proposing locally to see how it affects WooCommerce and other plugins that rely on the `after_password_reset` hook.

**Comment by mfs-mindsize** [comment]
@loranallensmith I'm working on another issue ATM. I'll look at this one next. If nothing else, I'm curious about the difference between the two logout methods prior to the hook we've been talking about. How many different ways can you log someone out? :) Hopefully, I'll soon find out.

**Comment by mfs-mindsize** [comment]
@loranallensmith - OK. I've had some time to look at this. You probably don't need all these details but I might as well share it anyway (for future reference). Perhaps someone else will chime in on the reasons for the various deltas. 

1. The WP core function reset_password() is pluggable [https://codex.wordpress.org/Pluggable_Functions](https://codex.wordpress.org/Pluggable_Functions) but WC's static method is not. Why? The impact(s)? I don't have those answers handy. But let's presume - until we hear otherwise - there's good reason for it. Maybe? ;)

2. The WP function has this line but WC does not. That's odd to me. What is it about WC that this isn't necessary?
`update_user_option( $user->ID, 'default_password_nag', false, true );`

3. WC's method has these lines (below) but WP does not.  Again, it's not clear to me why there should be a difference but again let's presume there's good reason for it. 
```
self::set_reset_password_cookie();

if ( ! apply_filters( 'woocommerce_disable_password_change_notification', false ) ) {
	wp_password_change_notification( $user );
}
```

4. **tl;dr What would I do? (Or what would I do if I were the plugin dev?)**
I'd stick to using:
`do_action( 'password_reset', $user, $new_pass );`

why? Because for all intents and purposes, in the case of sending emails, it's the same hook. That is, before or after are simply a matter of timing. If before happens - regardless of the success or failure of the wp_set_password() - so will after. I say this because there's no (if) conditional within the WP function that checks for the success or failure of wp_set_password(). In fact, that function doesn't even return a value. [https://developer.wordpress.org/reference/functions/wp_set_password/](https://developer.wordpress.org/reference/functions/wp_set_password/)

Put another way, this is **NOT** what the WC core function is doing:

```
$wp_result = wp_set_password( $new_pass, $user->ID );
if ( 'success' === $wp_result ){ 
    update_user_option( $user->ID, 'default_password_nag', false, true );
    do_action( 'after_password_reset', $user, $new_pass );
} else {
    do_action( 'after_password_reset_failure', $user, $new_pass );
}
```
For better or worse, since there is nothing looking at the success or failure, the two hooks are identical. Of course, there are use cases where the timing does matter. That you might not want to do Y unless you were using the after hook. I don't think that applies to these emails. 

5. WWID - Pt 2 - Along with using the common 'password_reset' hook, to avoid the dupe emails you've mentioned, the plugin simply needs to incorporate the conditional that's in the WC method:
```
if ( ! apply_filters( 'woocommerce_disable_password_change_notification', false ) ) {
	...whatever it is the plugin wants to do...
} 
```
It's likely the BNFWP plugin would also have to check for WC being installed and active. I can't say for sure. I've rambled on far too long already. But if that plugin has a public repo I can try to have a look.

Bottom line:
- I'm confident that it's not 100% necessary to modify core in order for BNFWP to work properly. 
- None the less, I'll knock out a PR so WC has the after action as well. But keep in mind, core's review and merge process could take some time.

**Comment by mfs-mindsize** [comment]
FYI - PR submitted https://github.com/woocommerce/woocommerce/pull/28938

**Comment by EngKhaledB** [comment]
Hi @loranallensmith @jackmcconnell,

WooCommerce is already hooking the customer password reset with this action: `woocommerce_customer_reset_password` on https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-form-handler.php#L1054
It fires the hook after a successful password reset. So, I see there is no need to define new hooks.

**Comment by jackmcconnell** [comment]
Thanks @EngKhaledB,
The problem I've had is knowing where the user registration occurred - did it happen through WooCommerce, or another plugin, or via the default WordPress registration form page?

I can test to see if WooCommerce is active on the site but I can't assume that all registrations come through one plugin. 
If you know of a way I can see where the user registration originated from, I may be able to use the hook you described when WC is installed and activated.
